### DISCUSSION TITLE
Link draw.io and ChatGPT

### DISCUSSION THREAD

**Discussion by alderg** [jgraph/drawio#4044] [General]
> [!NOTE]  
> **This discussion has been moved to https://github.com/jgraph/drawio/discussions/5387**

You can now enable a generic ChatGPT window which allows to create new diagrams, ask questions about the current diagram, selection or the app (help). The feature is experimental and requires a [ChatGPT API key](https://help.openai.com/en/articles/4936850-where-do-i-find-my-api-key). To enable the feature, use `gpt-api-key=[your-api-key]` as an URL parameter or specify `gptApiKey` in the configuration (see below). An optional `gpt-model=[model-name]` can also be specified. Possible values are `gpt-4` or `gpt-3.5-turbo` (default), depending on the API key.

![Screenshot 2023-12-15 at 13 05 27](https://github.com/jgraph/drawio/assets/921200/81eb6a47-0e99-43fd-8bc0-a71441f7dc10)

If an ChatGPT API key is present, the _Chat_ menu will be shown under the _Tags_ menu:

<img width="442" alt="Screenshot 2023-12-15 at 13 02 30" src="https://github.com/jgraph/drawio/assets/921200/d14cf2e9-a4e2-4886-81f7-295ed961f96c">

### Examples

Creating diagrams and summaries:

![8199d3a34f342f7e9dce8bf379828a8e](https://github.com/jgraph/drawio/assets/921200/0a522a1e-0624-42e0-86db-7c96c3a2e014)

Asking questions about cross-functional flowchart roles:

![285945477-9aa1d832-74e6-4f7f-ad9a-fadf9ce8d6a0](https://github.com/jgraph/drawio/assets/921200/3dc26f79-f280-45cb-a2f8-a7f4d84edc74)

Changing styles of vertices and edges:

![285945532-a8a1946a-3e1f-46a0-bc08-590af25dd9b7](https://github.com/jgraph/drawio/assets/921200/97f4bf01-015e-4984-84be-f40b28e222d8)

Describing and interpreting org charts:

![285945720-a9dfcdb0-e69c-480c-9da0-558f67e7fc47](https://github.com/jgraph/drawio/assets/921200/2d9a745b-26db-4bd7-ae48-536b440a0506)

Asking questions about class diagrams and multiplicities:

![285947153-d34649b6-4fe7-496c-969b-0a4c4148f477](https://github.com/jgraph/drawio/assets/921200/993e3346-5ad4-4108-b587-6ff19e3e5280)

Describing sequence diagrams:

![285953282-214e7215-e7e0-4f08-b920-454fc190e3f7](https://github.com/jgraph/drawio/assets/921200/c213d86e-ce90-4317-988c-0b0d56483074)

Limitations for number of tokens on large diagrams:

![285953335-9fd0971f-dfa2-4a9e-a0f0-63bae2addcca](https://github.com/jgraph/drawio/assets/921200/73aec2bd-50f6-4950-8ed9-33674d7fd0cb)

Asking questions about application functionality (help):

![285953629-3836aa05-c214-4d57-8727-d549858eaf2a](https://github.com/jgraph/drawio/assets/921200/48265061-ed53-43b1-8bc4-e5c5eb2d7289)

**Comment by JustSaX** [comment]
Is a specific version necessary? I'm using v23.0.2 on mac. An entered the following in configuration:

`{
  "gptApiKey": "xxxx"
}`

I don't have the option for chat in the view menu.

**Reply by davidjgraph** [reply]
the offline desktop app?

**Reply by JustSaX** [reply]
yes

**Reply by davidjgraph** [reply]
it's offline, meaning it doesn't connect to any online servers, as would be required to call chatgpt.

**Comment by thecopeeninja** [comment]
Can we use self hosted LLM models in configuration file?
```
{
  "enableChatGpt": true,
  "gptApiKey": "dummy-key",
  "gptModel": "TechxGenus/DeepSeek-Coder-V2-Lite-Instruct-AWQ",
  "gptUrl": "http://YOUR_TAILSCALE_GPU_IP:8000/v1/chat/completions"
}
```

**Reply by alderg** [reply]
It should work.

**Reply by thecopeeninja** [reply]
I have been trying for the past two weeks but I keep getting error in chat interface when using above custom configuration. 
Are you able to replicate it on your end? 

Also, the windows installed app of drawio editor doesn't support chat interface. Only web browser version supports it for now.

**Reply by davidjgraph** [reply]
1. What is returned in the network tab of the browser debugger for that call?
2. That statement is correct.

**Comment by hushoujier** [comment]
May I ask if it is possible to add a Content-Security-Policy for `https://api.siliconflow.cn` in content-src。
![图片](https://github.com/user-attachments/assets/8ef678f0-3929-494a-a019-465cec2e0899)

Thank you.

**Reply by davidjgraph** [reply]
Absoutely not.

**Comment by futugyou** [comment]
May I ask if it is possible to add a Content-Security-Policy for https://models.github.ai in content-src。
Same reason as the previous reply.

![image](https://github.com/user-attachments/assets/574cae28-3609-4523-8bc8-7ede759e1dfe)

**Reply by davidjgraph** [reply]
Absolutely not. If you want to allow XHR to some random site, host the project yourself and allow the CSP.

**Reply by futugyou** [reply]
Can't you see that this is provided by github, and definitely not the so-called 'some random site' you mentioned.
Your CSP list has `https://api.github.com`, which shows that you are not excluding github, and `https://models.github.ai` is provided by github.

**Reply by davidjgraph** [reply]
I don't care. We don't use the domain, we're not adding it.

**Reply by tfbroner** [reply]
> Can't you see that this is provided by github, and definitely not the so-called 'some random site' you mentioned. Your CSP list has `https://api.github.com`, which shows that you are not excluding github, and `https://models.github.ai` is provided by github.

You can't be serious. How can you ask JGraph to vouch for the content served on a remote server which is NOT under JGraph's control? It's absurd.

**Comment by mesbahworld** [comment]
**Because Open AI are not free and I want to add some opensource or free option but unable to get succeed.**

For **Gemini** they don't use bearer in the header and it seems drawio by default using bearer and that's why i am getting error.
  If i use **Ollama** and for test I did not use any authentication. so if i dont use **gptApiKey** the chat option not appear and if i use **dummy key** then the request failed.
  
**Gemini:**
```

curl --location 'https://generativelanguage.googleapis.com/v1beta/models/gemini-2.0-flash:generateContent' \
--header 'Content-Type: application/json' \
--header 'X-goog-api-key: MyG00g1e_T0KeN' \
--data '{
    "contents": [
      {
        "parts": [
          {
            "text": "Explain how AI works in a few words"
          }
        ]
      }
    ]
  }'

```
<img width="1498" height="913" alt="image" src="https://github.com/user-attachments/assets/ee10aa43-c3e9-4af0-845c-040f97549e61" />

**Ollama:** 
```

curl --location 'http://MY_IP:11434/api/chat' \
--header 'Content-Type: application/json' \
--data '{
    "model": "codellama:13b",
    "messages": [
      {
        "role": "user",
        "content": "Why is the sky blue?"
      }
    ]
  }'
  
```
  
  
<img width="1578" height="913" alt="image" src="https://github.com/user-attachments/assets/e0292a8b-6b63-4209-af75-a489fd6c5a10" />

**Reply by mesbahworld** [reply]
> Adding Custom header option would be nice to have other vendor 
or,
Let user choose the full **'key : value'** as user input instead of default **'bearer'** : key_user_input

**Reply by alderg** [reply]
See https://github.com/jgraph/drawio/discussions/5387
